emoji: Use some more informational source for generating mapping information.

HarshitOnGitHub opened the pull request:
Fixes: #3972.

timabbott commented:
Just took a quick look to get a feel for how you're doing this.  Re; the iamcal data, I think we might want to implement a system where `tools/provision` does a clone+checkout of the repo, and we just have to update the commit ID in use, rather than checking in the JSON.  But I'd wait to do the actual work of writing that provisioning tooling until we're confident we're happy with that data set (it's a requirement to merge, not to prototype and figure out how this should work!).

HarshitOnGitHub commented:
@timabbott @rishig I have pushed here a working prototype to showcase what iamcal's emoji.json is offering to us. It's just a working prototype not for review. Can you have a look!?

timabbott commented:
Cool, this seems pretty good; @rishig as our resident emoji expert, do you want to check it out as well?  You need to provision, and it's only in the compose emoji picker (not reactions).  

I did notice a few things:
* I think this breaks the symlink farm or something, because I tried sending some emoji, and got 404s.  Didn't investigate.
* We don't support skin tone modifiers in our markdown right now, so we should probably hide that category until we can implement it (and we might want to make it a toggle UI-wise anyway when we do).  
* `emoji_codes.js` is still only 68K uncompressed, so while we could probably optimize it, I don't think we strictly need to for now.

In `emoji_codes.js`, do we want the category mapping to be to emoji names?  I was kinda thinking we might want it to go like this:

* We have a bunch of codepoint -> (list_of_emoji_names) objects.
* Categories are represented as a map from categories to lists of codes.

but @rishig may have thoughts on the long-term structure.

timabbott commented:
I think it's reasonable for you to start working on a system for managing `emoji.json` from a Git ID in their repository.  

For the picker, I just noticed https://github.com/needim/wdt-emoji-bundle, which uses the iamcal data set (just saw this linked).  It could be worth seeing if integrating (or forking) that would be an easier path than making a nice picker from bootstrap popovers.

HarshitOnGitHub commented:
@timabbott The main problem with https://github.com/needim/wdt-emoji-bundle is that it currently has no support for custom emojis and also there are some glitches like text not showing while I type in the textbox etc.

![anim](https://cloud.githubusercontent.com/assets/16687990/23642982/d1ce9136-0324-11e7-8d46-af971969b4a5.gif)

timabbott commented:
Yeah, OK, sounds like it may be buggy.  But since it's MIT-licensed, we can probably borrow some of their CSS :)

rishig commented:
```quote
In emoji_codes.js, do we want the category mapping to be to emoji names? I was kinda thinking we might want it to go like this:

    We have a bunch of codepoint -> (list_of_emoji_names) objects.
    Categories are represented as a map from categories to lists of codes.

but @rishig may have thoughts on the long-term structure.
```

yeah, I think category -> list of unicode codepoints is probably good as a long term structure. 

We'll also need a category for zulip.png (the "unicode codepoint" there would be `zulip`) , and for any future custom zulip emoji. Let's just call that category "Zulip".

HarshitOnGitHub commented:
I am stuck on this PR from the past one and half week without much progress. With every change that I am making I am getting a feeling that our emoji system is getting more and more complex and fragile and maintaining it in the future will be a real pain. I think we need a more structured plan to fix our emoji system from the starting itself. @timabbott @rishig thoughts?

timabbott commented:
Well, I think once we have everything working well off the iamcal emoji data sprite sheets, I think it should be possible to remove most of the older code, and thus simplify the system.  The problem is that before doing that we'll need to migrate historical messages (or otherwise put in a backwards-compatibility system), but I think that should be manageable.

But until we're ready to do that, the system will definitely increase in complexity for a while.

timabbott commented:
What problems are you stuck on currently @HarshitOnGitHub?  I can start reviewing and merging pieces if that'd be helpful.

HarshitOnGitHub commented:
There is no specific problem as such but as I make some new change that change tends to break some other part of the code and due to the complexity it sometimes become really difficult to track if it's going to break something and most of the times it does and I need to add some kind of handcoded exceptional logic to fix it which I think will make it harder to make changes at a later stage.

timabbott commented:
Yeah, that's life working with a complicated system that doesn't have tests :).

I suspect we can mostly solve the long-term maintainability problem by writing tests for the places where we needed to hardcode an exception (e.g. Node tests that inspect `emoji_codes.js` and check that certain emoji map in a certain way).  I think those should be pretty quick to write.  

I think the medium-term plan should be for new messages just use the iamcal data set (and not even use the old emoji-map.json) and actually just use their sprite sheets directly as well (I guess we'll have to move the zulip emoji back to an independent image), since then the only outputs we need to produce are `emoji_codes.js` and sprite sheet CSS output pieces.  That seems like something we should be able to maintain pretty easily once we get there.

HarshitOnGitHub commented:
Yeah, I think removing our sprite generation code is going to reduce a lot of complexity.

rishig commented:
I definitely remember feeling overwhelmed by the complexity of the emoji system when I was dealing with it.

For backwards compatibility, is the symlink farm sufficient? Realm emoji are just external links anyway, right?

HarshitOnGitHub commented:
symlink farm is sufficient for maintaining the backward compatibility as long as we can generate it without errors. We are using `emoji_map.json` for generating the symlink farm and it differs from the iamcal's data set at a number of points. My last commit fixes just that. But I have a list of several emojis which I am not able to extract from iamcal's data set as they would require hand coded mapping and hence no symlinks are being generated for them at this stage.

rishig commented:
There may be a misunderstanding here. I think the proposal is to have two completely independent systems:
* System 1, only used for rendering old messages: symlink farm (generated by emoji_map.json), which point to png files extracted from the ttf files. No changes are needed here.
* System 2, for new messages: We use the iamcal data set, and the iamcal sprite sheets. The rendered emoji look something like `<span class=emoji-u1234>:simple_smile:</span>`, and a bunch of generated CSS causes this to appear like :slightly_smiling_face: rather than `:simple_smile:`.

So there should be no interaction between the old names and the new names, and no symlinks should be generated from the iamcal data. Does that make sense? We can also move this to chat.zulip.com, just at-mention me there if you start a thread.

HarshitOnGitHub commented:
@rishig 
>System 1, only used for rendering old messages: symlink farm (generated by emoji_map.json), which 
point to png files extracted from the ttf files. No changes are needed here.

The symlinks do not directly point to the png files extracted from the ttf file rather they to the png files in  out/unicode which are essentially copies of png files extracted from the ttf files after being renamed using a mapping generated from NotoColorEmoji.ttx and iamcal's dataset and there are some codepoints on which either emoji_map and iamcal's dataset don't agree or they are not bound to any short name in iamcal's dataset.

HarshitOnGitHub commented:
@timabbott @rishig I have made some changes to use the iamcal's sprite sheets. Currently the size of each emoji in emoji picker is 25px but since iamcal's sprite sheets don't offer a 25px sized sprite sheets, I have used 20px sprite sheet(32px is looking too big). Are they looking fine to you?
One thing that you might notice is that some images are missing and it is because of differences between `emoji_map.json` and iamcal's `emoji.json`. It will be fixed as soon as I deprecate the use of `emoji_map.json`(next commit).

timabbott commented:
We can use e.g. the 64px one and then scale it down with CSS (actually this is part of how we'll be fixing the perf issue with scrolling on Mac, since we need to do 18px or something for emoji reactions, and the current scaling strategy we're using doesn't work).

timabbott commented on tools/setup/emoji/download-emoji-data:
```diff
@@ -20,4 +20,5 @@ if [ "$current_commit" == "$commit" ]; then
 else
     echo "$commit" > "$commit_file_path"
     wget "https://raw.githubusercontent.com/iamcal/emoji-data/$commit/emoji.json" -O "$output_path/emoji.json"
+    wget "https://raw.githubusercontent.com/iamcal/emoji-data/$commit/sheet_google_20.png" -O "$output_path/sheet_google.png"
```
I think we probably want to keep the resolution in the filename.

timabbott commented on tools/setup/emoji/build_emoji:
```diff
@@ -239,6 +252,30 @@ def dump_emojis(cache_path):
+            emoji_classes.append(emoji_class % {"codepoint": emoji["unified"], })
+            emoji_positions += emoji_pos_info % {'codepoint': emoji['unified'],
+                                                 'pos_x': emoji["sheet_x"] * 20,
+                                                 'pos_y': emoji["sheet_y"] * 20,
```
Let's use percent, not pixels, for this; that is necessary for resolving the emoji scaling performance issue.

(If you play with Slack's CSS for their emoji, you can see an example of how this works).

HarshitOnGitHub commented:
Big change: Use sprite sheets to render emojis everywhere. Still to be tested thoroughly(especially with `realm_emojis`).
Next steps: Manual testing, fixing tests and adding categorization to pickers.

timabbott commented:
Nice!  Let's leave categorization off this PR -- the goal for now is just to do the exact same thing we're doing now with the new toolchain.

HarshitOnGitHub commented:
@timabbott @rishig This is ready for a review!!

rishig commented on tools/setup/emoji/download-emoji-data:
```diff
@@ -0,0 +1,23 @@
+set -e
+set -x
+
+output_path=static/third/emoji-data
```
Other places in the codebase seem to use capital letters for these types of variables

rishig commented on tools/setup/emoji/build_emoji:
```diff
@@ -239,6 +255,31 @@ def dump_emojis(cache_path):
+    sprite_sheet_google = os.path.join(EMOJI_DATA_PATH, 'sheet_google_32.png')
+    run(['cp', sprite_sheet_google, cache_path], shell=True)
+    emoji_classes = []
+    emoji_class = ".emoji_%(codepoint)s"
```
I'd maybe call this emoji_class_template (or similar), and similarlywith emoji_pos_info.

timabbott commented:
I played around with this, and it works, which is awesome :).

@HarshitOnGitHub presumably we could move this section of the sprite sheet CSS to just be attached to `span.emoji`, rather than needing to tag every single `.emoji-foo` with them, right ?

```
{
    display: inline-block;
    height: 20px;
    width: 20px;
    background-image: url('sheet_google_32.png');
    -webkit-background-size: 1025px 1025px;
    -moz-background-size: 1025px 1025px;
    background-size: 1025px 1025px;
    background-repeat: no-repeat;
}
```

And then perhaps we can use `img.emoji` for the backwards-compatibility CSS for existing messages?

That's an optimization we can do in a follow-up if needed; I think at this point we can work on trying to clean this up and merge this.

Re: the `_`s, I noticed in the sprite sheet CSS that emoji that have a longer unicode spelling look like this:

`.emoji_1F469-200D-2764-FE0F-200D-1F48B-200D-1F469`

Is that the motivation for the `_` delimiter?  I feel like it might still be better to stick with `-`, unless I'm missing something.

One bug I noticed: the `:zulip:` emoji doesn't appear in the typeahead autocomplete.

HarshitOnGitHub commented on tools/setup/emoji/emoji_setup_utils.py:
```diff
@@ -223,3 +223,19 @@ def emoji_names_for_picker(emoji_map):
+def generate_emoji_catalog(emoji_data):
+    # type: (List[Dict[Text, Any]]) -> Dict[str, List[str]]
+    sort_order = {} # type: Dict[str, int]
+    emoji_catalog = {} # type: Dict[str, List[str]]
```
There is some problem in stringifying a `defaultdict()`, so I am leaving it as it is.

HarshitOnGitHub commented:
@timabbott Addressed all the comments. Ready for a final sign-off. There is a known issue, i.e., adding zulip emoji as a reaction gives a traceback as zulip emoji is a fake realm emoji and present only in frontend. I would like to fix that issue as a followup.

rishig commented on tools/setup/emoji/emoji_setup_utils.py:
```diff
@@ -242,3 +238,7 @@ def generate_emoji_catalog(emoji_data):
         emoji_catalog[category].sort(key=lambda codepoint: sort_order[codepoint])
     return emoji_catalog
+
+def image_available_for(emoji):
```
I'd call this function `emoji_is_universal`, and the argument `emoji_dict` (or something other than `emoji`).

rishig commented on zerver/lib/bugdown/__init__.py:
```diff
@@ -678,32 +678,43 @@ def handleMatch(self, match):
+def make_emoji(codepoint, display_string):
+    # type: (Text, Text) -> Element
+    span = markdown.util.etree.Element('span')
+    span.set('class', 'emoji emoji-%s' % (codepoint,))
```
Can these `span`s also have an `alt`? To make it easier to implement the "don't show me any emoji" setting.

timabbott replied:
We can just hide the spans.  The "alt text" is the actual text in the span.  So I think we're good on accessibility.

timabbott commented:
> Search has strange behavior. E.g. try sending a message with :octopus:, and then searching for octopus in search. Sending a message with :zulip: and searching for zulip is similarly broken (though in a slightly different way).

I think searching for emoji has always been broken; see #3051.

rishig commented:
agreed, though now it is a bit worse, I think. Also wasn't sure if it was due to the changes in this PR to `content_highlighter`. But fine with this being a followup issue.

HarshitOnGitHub commented:
@rishig Thanks for the pretty detailed review!! Will start cleaning up this from tomorrow.

rishig commented on zerver/lib/bugdown/__init__.py:
```diff
@@ -678,32 +678,43 @@ def handleMatch(self, match):
     elt.set("title", display_string)
     return elt
 
+def to_codepoint(value):
```
I would rename this to something like `unicode_emoji_to_codepoint`, and also use something more evocative than `value`. In the best case, someone reading just the `def` line and the mypy arguments can have a good guess at what the function does.

rishig commented on zerver/lib/bugdown/__init__.py:
```diff
@@ -678,32 +678,43 @@ def handleMatch(self, match):
+def to_codepoint(value):
+    # type: (Text) -> Text
+    surrogate_pairs = []
+    for a in six.text_type(value):
```
Can this be
`surrogate_pairs = [hex(ord(a))[2:] for a in six.text_type(value)]`?
Also, similar comment to above, `a` could be replaced with a more descriptive name.

rishig commented:
(Added two more small comments :).)

HarshitOnGitHub commented:
@rishig Took care of all the comments except the two refactoring and skin tone modifier comments. Would like to do them as follow-ups.

rishig commented:
cool, sounds good!

rishig commented:
Do you know why things like `:football:` aren't showing up in the typeahead here?

HarshitOnGitHub commented:
@rishig This is because it is one of the blacklisted names.

rishig commented:
ah, `emoji_names_for_picker` and anything else that refered to the old gemoji names should be moved over to the new iamcal names. For now I think it's fine to use the iamcal shortcode field as the emoji picker name. Also as a follow-up we should definitely remove all the now-dead code that was there for the gemoji set.

HarshitOnGitHub commented:
Done. :+1:

rishig commented on tools/setup/emoji/emoji_setup_utils.py:
```diff
@@ -15,115 +15,6 @@
-
 # A handcoded list of popular emoji names. This is useful for cases
 # where the two names are very different, users might reasonably type
 # either name and be surprised when they can't find the relevant emoji.
```
I see, I didn't realize we already had the ability for multiple names in the autocomplete. In that case I think we should just include all the short_codes programatically. And maybe blacklist 'shit'?

In case you're curious, here is the list:

[(u'phone', [u'phone', u'telephone']),
 (u'boat', [u'boat', u'sailboat']),
 (u'email', [u'email', u'envelope']),
 (u'hand', [u'hand', u'raised_hand']),
 (u'exclamation', [u'exclamation', u'heavy_exclamation_mark']),
 (u'moon', [u'moon', u'waxing_gibbous_moon']),
 (u'mostly_sunny', [u'mostly_sunny', u'sun_small_cloud']),
 (u'barely_sunny', [u'barely_sunny', u'sun_behind_cloud']),
 (u'partly_sunny_rain', [u'partly_sunny_rain', u'sun_behind_rain_cloud']),
 (u'lightning', [u'lightning', u'lightning_cloud']),
 (u'tornado', [u'tornado', u'tornado_cloud']),
 (u'runner', [u'runner', u'running']),
 (u'izakaya_lantern', [u'izakaya_lantern', u'lantern']),
 (u'bee', [u'bee', u'honeybee']),
 (u'dolphin', [u'dolphin', u'flipper']),
 (u'feet', [u'feet', u'paw_prints']),
 (u'facepunch', [u'facepunch', u'punch']),
 (u'+1', [u'+1', u'thumbsup']),
 (u'-1', [u'-1', u'thumbsdown']),
 (u'shirt', [u'shirt', u'tshirt']),
 (u'mans_shoe', [u'mans_shoe', u'shoe']),
 (u'family', [u'family', u'man-woman-boy']),
 (u'couple', [u'couple', u'man_and_woman_holding_hands']),
 (u'boom', [u'boom', u'collision']),
 (u'hankey', [u'hankey', u'poop', u'shit']),
 (u'book', [u'book', u'open_book']),
 (u'memo', [u'memo', u'pencil']),
 (u'hocho', [u'hocho', u'knife']),
 (u'middle_finger',
  [u'middle_finger', u'reversed_hand_with_middle_finger_extended']),
 (u'laughing', [u'laughing', u'satisfied']),
 (u'car', [u'car', u'red_car']),
 (u'the_horns', [u'the_horns', u'sign_of_the_horns']),
 (u'flag-cn', [u'flag-cn', u'cn']),
 (u'flag-de', [u'flag-de', u'de']),
 (u'flag-es', [u'flag-es', u'es']),
 (u'flag-fr', [u'flag-fr', u'fr']),
 (u'flag-gb', [u'flag-gb', u'gb', u'uk']),
 (u'flag-it', [u'flag-it', u'it']),
 (u'flag-jp', [u'flag-jp', u'jp']),
 (u'flag-kr', [u'flag-kr', u'kr']),
 (u'flag-ru', [u'flag-ru', u'ru']),
 (u'flag-us', [u'flag-us', u'us'])]

rishig commented:
Left a comment and then deleted it, in case you got a confusing notification. 
Actually let's not merge the dead code removal commit just yet; after the rest of this is merged, I'll go through and add the alternate names and remove the old name code in a single commit. (It seems like the new code already supports the ability to have multiple names in the autocomplete.)

HarshitOnGitHub commented:
@rishig You are right, we don't support multiple names for the same emoji in autocomplete yet. I had added that list as a workaround just for displaying some of the most used emoji names by duplicating them in the autocomplete but it led to a duplication in the emoji pickers as well so I have removed that list for now. So until the autocomplete issue is fixed, two or three names like `thumbsup`, `thumbsdown` etc will not be available in the autocomplete although their alternatives like `+1`, `-1` etc are available in the autocomplete.

rishig commented:
Just as a note, this is probably the highest priority follow-up, since I don't think we can release 1.6 without it.
Thanks for your work on this!

HarshitOnGitHub commented:
Ah, one more issue! Some reactions on old messages like 'simple_smile' will not work as we store name for reactions in database and those names are no longer available in `emoji_name_to_css_class` mapping. Looks like we will need a tool to convert the old reactions in database to newer ones. Thoughts?

rishig commented:
I haven't looked, but probably we should just do a migration to convert Reaction.emoji_name to Reaction.codepoint.

HarshitOnGitHub commented:
Yeah migration will do the work! :)

timabbott commented:
You mean a migration to update all the historical messages?

rishig commented:
A migration just to update the `Reaction` table. The reactions aren't stored in the `Message`/`UserMessage` table so I think we're good.

timabbott commented:
Ahh, yeah, that should be totally doable.

timabbott commented:
One thing to be careful about is that we do support realm emoji in reactions, which won't have a codepoint.

timabbott commented:
Thanks for doing this @HarshitOnGitHub!  

I just reviewed and merged everything but the "Interrelated emoji changes" commit (actually merged one hunk of that as well that was needed to merge some of the downloading pieces of #4383 without too much merge conflict trouble.  

So no user-facing functional changes have been merged just yet -- I guess we'll need to add the reactions migration before we can do that, but if you rebase, this branch should now be a lot smaller to maintain and re-review :).

HarshitOnGitHub commented:
Thanks a ton @timabbott!! My semester exams approaching so I am getting a bit less time to work on this but trying my best to complete it soon. :)

timabbott commented:
A major question to decide before doing the reactions migration is what our strategy is there -- are we going to collapse all the reactions with the same image onto the codepoint, so that there is no record of what the user "actually" selected, just the image?  If we're comfortable with that model, then I think it's (mostly) clear how to do the migration.

@rishig your thoughts on this would be appreciated.

timabbott commented:
@HarshitOnGitHub if you have time to at least rebase this, that would be helpful for letting others help get this out while you're busy with exams :).

HarshitOnGitHub commented:
Sure!

HarshitOnGitHub commented:
>are we going to collapse all the reactions with the same image onto the codepoint, so that there is no record of what the user "actually" selected, just the image?

I think we can keep the original emoji name that the user selected in a separate field besides having emoji codepoint in another field and use the original emoji name to render it as text and the codepoint to render it as an image. Does that make sense?

timabbott commented:
That idea makes sense to me; I'm interested in @rishig's thoughts though, since he's thought about this stuff a lot more than I have :).

rishig commented:
hm, I think it's probably not worth saving the emoji_name. One issue is what happens if one person reacts with `thumbsup`, and another with `+1`. We probably want that to be visually just one emoji. 

This is what slack does, fwiw. If you send an emoji in a message, slack will preserve what you typed. If you do an emoji reaction, slack converts it to its canonical shortname.

timabbott commented:
I think we can go with that; it's less work and we can always change it later.

rishig commented:
A quick note on doing the migration: You'll probably have to write it by hand. Instead of renaming the emoji_name column, I would do it as: first add the codepoint column (or whatever we want to call it), then fill it in, and then delete the emoji_name column.

Also, normally it wouldn't make sense to refer to outside files in a migration, but I think here using emoji_map.json should be fine, since we're treating that as a fixed, unchanging file that we're keeping around ~forever.

timabbott commented:
@HarshitOnGitHub I've merged a bunch of chunks of this into master; the remainder is on the timabbott/emoji-infra branch.

timabbott commented:
So I did a bit more work on this on my `emoji-infra` branch.  Here's a status update:
* There's some CSS sizing issues that we'll need to resolve in the emoji picker, but otherwise things look good in the current version as long as you don't have to migrate old reactions forward.
* I did a draft of step 1 of the reactions migration, namely translating emoji_name values to codepoints (while at it, I think we can flag in the DB which are realm emoji).  It's going to be some work to fully finish it.
* The reactions migration will require us to send a codepoint-to-canonical-name data set to the frontend, which we haven't currently planned.  We could alternatively defer that migration if we send to the frontend a old-emoji-name-to-codepoint mapping data set, which I'm increasingly thinking might be a better near-term solution for unblocking the core things we want to do here.  We'll need to keep that data set around ~forever in order for the Reactions migration to work anyway, so my thought is we can just check in a fixed JSON file with the data (and then we won't need to keep the code around), accessing that file for the migration code and (if we go that route) using the file as a data source for a temporary JS translation layer for reactions, as I described.

What are your thoughts on that last question @HarshitOnGitHub @rishig?

HarshitOnGitHub commented:
@timabbott Idea sounds good to me.

timabbott commented:
I just merged a few more small but important pieces of this branch; the end result is that we're now using the Iamcal sprite sheets (and the code for the old Zulip emoji sprite sheets is gone).  We should now be able to fix #4660 directly by using the CSS options afforded by our own sprite sheet.

There are a few major changes remaining for the rest of this branch:
(1) Switching emoji in messages to use the sprite sheets instead of individual images, for new emoji.  I have a commit staged that will do this; merging it is blocked on making sure the mobile apps will be able to handle the new format.
(2) Saving away a final copy of the old emoji name -> codepoint mapping that we can use for database migrations in the future.
(3) Adding to `emoji_codes.js` (or generating from it) a codepoint -> canonical name mapping that we can use for display purposes (e.g. for displaying a human-readable name for reactions).
(4) Switching reactions to store the codepoint, not the short names, using the above for backwards-compatibility.
(5) Consider the reactions API endpoints for adding/removing them to accept the codepoints?  I'm not sure whether we want to do that,.
(6) Switching to using the short names database from iamcal, rather than our historical short names -> codepoint mapping, which is basically the final remaining piece of this branch.

Huge thanks for your work on this so far @HarshitOnGitHub!  

I think (3) is the main piece of this that nobody has worked on yet, if you have a bit of time to help!

timabbott commented:
@HarshitOnGitHub can you rebase this to resolve the merge conflicts with @showell's recent reactions refactor?  Thanks!

HarshitOnGitHub commented:
This PR has been split into two PRs:
1: This one(aka emoji-web PR) which is blocked due to changes required in mobile apps.
2: Reaction migration PR(#5141) which is independent of this and can be merged now.

HarshitOnGitHub commented:
Just a note, removed the hacks used while preparing for zulip 1.6 release. This will require some more work once the reaction migration pr gets merged.

timabbott commented:
@HarshitOnGitHub if you get a chance, it'd be awesome to update this branch at some point.  But I think more important is closing out the reactions migration project, which I think is basically ready.

timabbott commented:
@HarshitOnGitHub now that the reactions migration is done, I think we should pick this up again! :)

HarshitOnGitHub commented:
@timabbott @rishig Since it was almost impossible to rebase the old branch I have redone this from scratch and this seems to be ready for a review. This apart from switching us to use sprite sheets also fixes a longstanding issue of number emojis being not searchable in emoji picker ever since we introduced emoji aliases. Also includes some fixes to notifications.py so that we don't break emoji rendering there.
Sorry for a single large commit but it was necessary to maintain commit atomicity.

P.S.: The test suite that I had added for relative URLs in digest emails is working well. It prevented me from introducing a bug related to zulip emoji.

timabbott commented on zerver/lib/notifications.py:
```diff
@@ -110,6 +109,18 @@ def relative_to_full_url(base_url, content):
+        content)
+    content = content.replace(' class="emoji"', ' style="height: 20px;"')
+
+    return content
```
Just a comment: I think we may want to eventually migrate this code path to use xml.ElementTree for doing the rewriting; that would likely be a lot more robust than the regular expressions we've been doing.

But it's not a regression, so OK for now :)

HarshitOnGitHub replied:
Yeah I was also thinking of replacing these regex with lxml while writing code for replacing spans with img tags.

timabbott commented:
It's great that the "big commit" is only ~100LOC, much of it changes in tests.  I just looked through it, and the first commit at least looks good.  I'll plan to test-deploy to chat.zulip.org today.

HarshitOnGitHub commented:
>It's great that the "big commit" is only ~100LOC, much of it changes in tests.

Haha, doing it took more than I had expected. Will it make sense to make a tool for automatically updating `markdown_test_cases.json` fixtures once user verifies that changes are actually correct.

timabbott commented:
I've for a while wanted a pretty-printer at least for the diff when you get a failure in `markdown_test_cases` failures, since it can be really hard to read as it is.  I think there's an issue for it.

timabbott commented:
A few small issues:
* In the typeahead, the emoji seem like they're a pixel too low.
![image](https://user-images.githubusercontent.com/2746074/30991993-610e0966-a45c-11e7-9907-52df8262a5c5.png)
* Also, `typeahead_helper.js` is missing complete node test coverage and thus failing CI.
* When `emojis_by_unicode` is removed, you forgot to remove the definition.
* We should bump PROVISION_VERSION to 10.0 somewhere in this branch, since we'll need `build_emoji` to be rerun when switching branches in either direction.

Also, can you summarize the impact expected from changing `name_to_codepoint` to be based on iamcal?  I want to make sure we have thought through everything.  Potential issues to think about:
* Existing emoji in messages.  This should be OK, since we don't access the emoji libraries at all when interacting with these
* Emoji reactions.  This should be OK if all of the code for what emoji to show is based on using the `emoji_code`, and the `emoji_name` is just used to print a name.  I'm not sure we have that quite yet; at least, the `unique_together` on the data model is still tied to `emoji_name`.

Can we split the syntax change in the sprite sheet from that without much effort?  If I'm not mistaken, 90% of the code in the first commit is the sprite sheet switch, and I'd love to be able to merge that piece right away (and get feedback) before confronting the name change problem, which is a lot more risky.

rishig commented:
I thought the name change was only for new codepoints? I.e. if a codepoint has a name in the current system, it's keeping its name till a bigger naming overhaul that should happen in a couple of weeks.

HarshitOnGitHub commented:
> Also, typeahead_helper.js is missing complete node test coverage and thus failing CI.

I had fixed it, may be forgot to push the new version, will push it now.

>When emojis_by_unicode is removed, you forgot to remove the definition.

Oops will fix it.

>We should bump PROVISION_VERSION to 10.0 somewhere in this branch, since we'll need build_emoji to be rerun when switching branches in either direction.

Yeah I had realized later and added a separate commit to bump the provision version while fixing the typeahead coverage issue, will push now.

Also there are no name changes in this PR. I have used the same names as we were using earlier.

HarshitOnGitHub commented:
>Also, can you summarize the impact expected from changing name_to_codepoint to be based on iamcal? I want to make sure we have thought through everything. Potential issues to think about:
>
> Existing emoji in messages. This should be OK, since we don't access the emoji libraries at all when interacting with these
Emoji reactions. This should be OK if all of the code for what emoji to show is based on using the emoji_code, and the emoji_name is just used to print a name. I'm not sure we have that quite yet; at least, the unique_together on the data model is still tied to emoji_name.

Another thing to think about it is notifications which we have already fixed. For emoji reactions we are already using the iamcal's version of `name_to_codepoint` with some hacky changes that I just removed in this PR.

HarshitOnGitHub commented:
Actually changing the `name_to_codepoint` to be iamcal based is going solve a number of problems. Till now there was an inconsistency in `name_to_codepoint` and `codepoint_to_name` data structures about which I had to remember everytime I made a change but both of them are now based on iamcal.

timabbott commented:
Merged, thanks @HarshitOnGitHub!  This has been one of the more epic projects in Zulip; huge thanks for all your work to make this happen!
